### PR TITLE
Feature:Lesson11

### PR DESCRIPTION
**Pull request by nyattsu72**
[課題11](https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#%E6%93%AC%E4%BC%BC%E7%9A%84%E3%81%ABapi%E3%82%92%E6%89%B1%E3%81%86)

[CodeSandBox](https://codesandbox.io/s/red-https-16l2g?file=/js/main.js)
Latest commit : [84f278e](https://github.com/nyattsu72/hanson-frontend-cource/pull/14/commits/84f278e600af7309dc3820b19652efd6cb1df830)

## Places you want to check
・I'm worried about the foreach array processing code, so please confirm.
・ I added a setTimeout of 3 seconds before the fetch process, but I'm worried if this code is okay.

### CODE REVIEW

File: lesson11/js/main.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    })
+  });
+
+  menuContainer.appendChild(menuLists).appendChild(fragment);
+}
+
+async function fetchMenuContents(){
+  showLoadingImg();
+  await new Promise((resolve) => setTimeout(resolve, 3000));
```
Comment (nyattsu72) [inline lesson11/js/main.js]: Please check this implementation.
I added a setTimeout of 3 seconds before the fetch process, but I'm worried if this code is okay.

File: lesson11/js/main.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  loadingImg.classList.add("el_loadingImg");
+
+  loading.appendChild(loadingImg);
+}
+
+function createMenuList(values) {
+  const fragment = document.createDocumentFragment();
+
+  Object.keys(values).forEach((key) => {
+    values[key].forEach(value => {
```
Comment (nyattsu72) [inline lesson11/js/main.js]: Please check this code.
I'm worried about the foreach array processing code, so please confirm.

File: lesson11/js/main.js (hunk 1 above)
Comment (ghost) [reply lesson11/js/main.js]: I reprint the comments I received from the doctor when he reviewed my work.

> setTimeoutはこの課題では不要です。setTimeoutは以前の課題までfetchしている間待つために擬似でやっていました。今はjsonデータを実際にリクエストして取っているのでここは不要です

So, there is no problem deleting this function.

File: lesson11/js/main.js (hunk 2 above)
Comment (443n0511) [reply lesson11/js/main.js:18]: [IMO](https://codesandbox.io/s/hidden-river-39jjb?file=/js/main.js)
I used the destructuring assignment (分割代入)
You can look into it if you want.😄

Comment (ghost) [reply lesson11/js/main.js]: [IMO](https://codesandbox.io/s/dazzling-architecture-yk1o3?file=/js/main.js)
I tried using Destructuring assignment(分割代入)
You can look into it if you want😄

It took me a long time to check the operation because of the bad chrome.
I'm sorry I've kept you waiting.😥

File: lesson11/js/main.js (hunk 1 above)
Comment (nyattsu72) [reply lesson11/js/main.js]: Thank you for sharing.

I removed setTimeout with this commit.

[6786034](https://github.com/nyattsu72/hanson-frontend-cource/pull/14/commits/67860346f836519c6ec6ed9a3360b63cb2436d95)

File: lesson11/js/main.js (hunk 2 above)
Comment (nyattsu72) [reply lesson11/js/main.js]: Thank you for review.

I know Destructuring assignment for the first time.
I tried using Destructuring assignment and was able to write the code simply.
Using forEach twice made it harder to read which values ​​were passed, so the code was easier to read.

I will try more array operations using Destructuring assignment.

[d8c39d9](https://github.com/nyattsu72/hanson-frontend-cource/pull/14/commits/d8c39d9ffd35b34182f09dd846778c8f08dedb8c)

**Review by ghost — APPROVED** [review]

File: lesson11/js/main.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+  const menuDataURL = 'https://jsondata.okiba.me/v1/json/Aticp210816214718';
+  const response = await fetch(menuDataURL);
+  
+  if(response.status === 200){
+    return await response.json();
+
+  }else{
+   console.error("Could not get the value");
+  }
```
Comment (shouyamamoto) [inline lesson11/js/main.js]: Try to use try-catch instead of if-else!

File: lesson11/js/main.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+}
+
+async function callMenuContents() {
+  try {
+    const menuContents = await fetchMenuContents();
+    createMenuList(menuContents);
+
```
Comment (shouyamamoto) [inline lesson11/js/main.js]: I didn't want the line breaks in each of these processes.
The line break is a single line, and the line break between functions is also a single line, so I felt that it could make it difficult to see where the break is.

If the process is so complicated that it's hard to see, or if you want to pause for a while, you may want to add a new line.
Otherwise, I think it's fine without line breaks!

File: lesson11/js/main.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+async function fetchMenuContents(){
+  showLoadingImg();
+
+  const menuDataURL = 'https://jsondata.okiba.me/v1/json/Aticp210816214718';
+  const response = await fetch(menuDataURL);
+  
+  if(response.status === 200){
+    return await response.json();
```
Comment (shouyamamoto) [inline lesson11/js/main.js]: Since we are using `return response.json` here, we need to use `createMenuList({data})`.
If we take one more step, we can avoid the `Destructuring assignment`.
I'm sure it's easier than you think, so once you figure out how to do it, please try to figure it out!

Comment (nyattsu72) [reply lesson11/js/main.js]: Thank you for review!

I understand that response.json () returns an javascript object, so I can use the usual array specification.
Thanks to that, I was able to make the code very simple.

[b5a3c2e](https://github.com/nyattsu72/hanson-frontend-cource/pull/14/commits/b5a3c2e55af358b61f64737b02fbd9492456b832)

File: lesson11/js/main.js (hunk 4 above)
Comment (nyattsu72) [reply lesson11/js/main.js]: Thank you for review!
I removed the line breaks.

[b5a3c2e](https://github.com/nyattsu72/hanson-frontend-cource/pull/14/commits/b5a3c2e55af358b61f64737b02fbd9492456b832)

File: lesson11/js/main.js (hunk 3 above)
Comment (nyattsu72) [reply lesson11/js/main.js]: Thank you for review!

I rewrote it to try-catch.
However, I'm worried about how to write the catch error code.
I would like to seek your advice on this matter.
Do I have to write an if statement with try here and pass the status code in case of an error?
I would appreciate your help.

[b5a3c2e](https://github.com/nyattsu72/hanson-frontend-cource/pull/14/commits/b5a3c2e55af358b61f64737b02fbd9492456b832)

File: lesson11/js/main.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    });
+
+  menuContainer.appendChild(menuLists).appendChild(fragment);
+}
+
+async function fetchMenuContents(){
+  showLoadingImg();
+
+  const menuDataURL = 'https://jsondata.okiba.me/v1/json/Aticp210816214718';
```
[the remark this answers is not in the document]
Comment (nyattsu72) [reply lesson11/js/main.js]: Thank you for review!

I capitalized the menuDataURL.
Also, menuDataURL is out of the function scope.

> if you want to run fetchMenuContents again, you will have to define the menuDataURL.
I didn't really care about it, so I'll be careful!

[bf8afa2](https://github.com/nyattsu72/hanson-frontend-cource/pull/14/commits/bf8afa26d14d992ec19c62853f323bb82662d4be)

**Comment by shouyamamoto** [comment]
I'm sorry, but there are a lot of fixes in one commit, and it's hard to follow the fix code.
Can you please update the codesandbox for me? Thank you very much.
Also, creating one commit for each fix makes it easier to check the code, so please take a look at that! : )

**Comment by nyattsu72** [comment]
> I'm sorry, but there are a lot of fixes in one commit, and it's hard to follow the fix code.
> Can you please update the codesandbox for me? Thank you very much.
> Also, creating one commit for each fix makes it easier to check the code, so please take a look at that! : )

I'm sorry, I forgot to update Codes and Box.
[CodeSandBox](https://codesandbox.io/s/red-https-16l2g?file=/js/main.js)
I updated it, so
Thank you for checking.

File: lesson11/js/main.js (hunk 3 above)
Comment (shouyamamoto) [reply lesson11/js/main.js]:
```
  try {
    const response = await fetch(MENU_DATA_URL);
    const json = await response.json();
    return json.data;
  } catch {
    console.error("error： HTTP error, status =" + response.status);
  }
```

It is done this way, but I think you are getting an error as `response` is `undefined`.
The reason for this is that the response is in the `try block`, and from the `catch block`, the `response` is out of scope.
Therefore, it is not clear at this point whether there is a way to pass the status code here, and I don't think it is that important.
I don't think it is that important, because the status code can be checked by the validation tool.
I think once we have it, we can just write a string that tells us that an error has occurred.

Comment (nyattsu72) [reply lesson11/js/main.js]: Thank you for review.

> It is done this way, but I think you are getting an error as response is undefined.

I understand that the constant responsen is out of scope for catch.

> I don't think it is that important, because the status code can be checked by the validation tool.

As Sho-san said, I didn't think I needed a status code here, so I made it simple to display the text.

 [84f278e](https://github.com/nyattsu72/hanson-frontend-cource/pull/14/commits/84f278e600af7309dc3820b19652efd6cb1df830)

**Review by shouyamamoto — APPROVED** [review]
